people-and-planet-ai/weather-forecasting/create_dataset.py

```diff
@@ -0,0 +1,233 @@
+) -> tuple:
+    """Gets an (inputs, labels) training example.
+
+    Args:
+        date: The date of interest.
+        point: A (longitude, latitude) coordinate.
+        patch_size: Size in pixels of the surrounding square patch.
+
+    Returns: An (inputs, labels) pair of NumPy arrays.
+    """
+    data.ee_init()
+    return (
+        data.get_inputs_patch(date, point, patch_size),
+        data.get_labels_patch(date, point, patch_size),
+    )
+
+
+def try_get_example(date: datetime, point: tuple) -> Iterable[tuple]:
+    """Wrapper over `get_training_examples` that allows it to simply log errors instead of crashing."""
+    try:
+        yield get_training_example(date, point)
+    except (requests.exceptions.HTTPError, ee.ee_exception.EEException) as e:
+        logging.error(f"🛑 failed to get example: {date} {point}")
+        logging.exception(e)
+
+
+def write_npz(batch: list[tuple[np.ndarray, np.ndarray]], data_path: str) -> str:
+    """Writes an (inputs, labels) batch into a compressed NumPy file.
+
+    Args:
+        batch: Batch of (inputs, labels) pairs of NumPy arrays.
+        data_path: Directory path to save files to.
+
+    Returns: The filename of the data file.
+    """
+    filename = FileSystems.join(data_path, f"{uuid.uuid4()}.npz")
+    with FileSystems.create(filename) as f:
+        inputs = [x for (x, _) in batch]
+        labels = [y for (_, y) in batch]
+        np.savez_compressed(f, inputs=inputs, labels=labels)
+    logging.info(filename)
+    return filename
+
+
+def run(
+    data_path: str,
+    num_dates: int = NUM_DATES,
+    num_bins: int = NUM_BINS,
+    max_requests: int = MAX_REQUESTS,
+    min_batch_size: int = MIN_BATCH_SIZE,
+    beam_args: Optional[List[str]] = None,
+) -> None:
+    """Runs an Apache Beam pipeline to create a dataset.
+
+    This fetches data from Earth Engine and writes compressed NumPy files.
+    We use `max_requests` to limit the number of concurrent requests to Earth Engine
+    to avoid quota issues. You can request for an increas of quota if you need it.
+
+    Args:
+        data_path: Directory path to save the data files.
+        num_dates: Number of dates to extract data points from.
+        num_bins: Number of bins to bucketize values.
+        max_requests: Limit the number of concurrent requests to Earth Engine.
+        min_batch_size: Minimum number of examples to write per data file.
+        beam_args: Apache Beam command line arguments to parse as pipeline options.
+    """
+    random_dates = [
+        START_DATE + (END_DATE - START_DATE) * random.random() for _ in range(num_dates)
+    ]
+
+    beam_options = PipelineOptions(
+        beam_args,
+        save_main_session=True,
+        setup_file="./setup.py",
+        max_num_workers=max_requests,  # distributed runners
+        direct_num_workers=max(max_requests, MAX_REQUESTS),  # direct runner
+    )
+    with beam.Pipeline(options=beam_options) as pipeline:
```

Comment:
That's awesome! I also think the Beam team made a great use of the `|` operator, just like unix pipes, but it can also confusing for some people. I personally like it :)